### PR TITLE
player: note authoring

### PR DESCRIPTION
**Pull request by vojtaholik**
[➽ live preview](https://skill-starter-turbo-6fw0rtb8k-skillrecordings.vercel.app/video/supabase-create-a-new-supabase-project)

https://user-images.githubusercontent.com/25487857/149927084-7887a81b-ac65-4d99-bf57-4d302e52f3a3.mp4

screenshot

gif

### CODE REVIEW

File: apps/next-product-starter/src/pages/api/lessons/notes/[slug].ts [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    const contact_id = req.query.contact_id
+    const {data: userNotes} = await loadUserNotesForResource(
+      req.query.slug as string,
+      contact_id as string,
+    )
+    const staffNotes = await loadStaffNotesForResource(
+      req.query.staff_notes_url as string,
+    )
+
+    const notes = await convertNotes(userNotes, staffNotes)
```
Comment (vojtaholik) [inline apps/next-product-starter/src/pages/api/lessons/notes/[slug].ts:66]: getting staff notes + user notes separately and converting them to VTT

File: apps/next-product-starter/src/pages/api/lessons/notes/[slug].ts [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+  contactId?: string,
+) => {
+  // all notes for the specific user
+  // all public notes
+  if (supabase) {
+    const {data, error} = await supabase
+      .from(process.env.RESOURCE_NOTES_TABLE_NAME || 'resource_notes')
+      .select()
+      .eq('resource_id', lessonSlug)
+      .or(`state.eq.published${contactId ? `,user_id.eq.${contactId}` : ``}`)
```
Comment (vojtaholik) [inline apps/next-product-starter/src/pages/api/lessons/notes/[slug].ts:94]: most of this is taken over from egghead-next and just works

File: apps/next-product-starter/src/pages/video/[slug].tsx [hunk 3]
```diff
@@ -208,20 +240,45 @@ const VideoCuesList: React.FC<any> = () => {
...
+            (context: VideoStateContext, _event: VideoEvent) => async () => {
+              context.viewer = VIEWER[1] as any
+            },
+        }}
         actions={{
           onVideoEnded: (_context: any, _event: any) => {
             console.log('do stuff')
           },
         }}
       >
```
Comment (vojtaholik) [inline apps/next-product-starter/src/pages/video/[slug].tsx:259]: `addCueNote` is a service passed down from product level which lets us make state transitions based on its result.

`loadViewer` is using hardcoded data at the moment as I needed way to test and demonstrate logic based on `contact_id`

File: packages/player/src/components/cue-form.tsx [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+import * as React from 'react'
+import {useVideo} from '../context/video-context'
+import {
+  selectCurrentTime,
+  selectIsSubmittingCueNote,
+  selectCueNoteVisibility,
+  selectViewer,
+} from '../selectors'
+
+export const CueForm: React.FC = () => {
```
Comment (vojtaholik) [inline packages/player/src/components/cue-form.tsx:11]: this is the form ui for authoring notes

File: packages/player/src/machines/video-machine.ts [hunk 5]
```diff
@@ -51,262 +71,402 @@ export interface VideoStateContext {
...
+        initial: 'focused',
+        on: {
+          SUBMITTED: {
+            target: 'taking_note.submitting',
+            actions: [
+              assign({
+                isSubmittingCueNote: (_context, _event) => true,
+                metadataTracks: (context, event) => {
+                  context.metadataTracks[0].addCue(event.cue)
+                  return context.metadataTracks
```
Comment (vojtaholik) [inline packages/player/src/machines/video-machine.ts:389]: we can simply `addCue` to current notes metadata track which visually updates them right away and then continue on making request to a server.

File: packages/player/src/machines/video-machine.ts [hunk 6]
```diff
@@ -51,262 +71,402 @@ export interface VideoStateContext {
...
+            on: {PLAY: 'playing'},
+          },
         },
       },
-      states: {
-        paused: {
-          on: {
-            PLAY: {
-              target: 'playing',
+      taking_note: {
```
Comment (vojtaholik) [inline packages/player/src/machines/video-machine.ts:379]: in order to get the authoring notes flow just right UX wise, I needed to distinguish between focused, typing, and submitting states.  state machine with `taking_note` state works a little like this:

state machine gif

File: packages/player/src/styles/scss/components/cue-form.scss [hunk 7]
```diff
@@ -0,0 +1,86 @@
+.cueplayer-react-cue-form {
```
Comment (vojtaholik) [inline packages/player/src/styles/scss/components/cue-form.scss:1]: this is where 💅 happens for cue form.

File: packages/player/src/components/shortcut.tsx [hunk 8]
```diff
@@ -92,186 +105,204 @@ export const Shortcut: React.FC<ShortcutProps> = ({
...
+          {
+            keyCode: 67, // c
+            handle: () => {
+              // focus the cue note input
+              cueFormElem?.input?.focus()
+              // start note taking session
+              videoService.send({type: 'TAKE_NOTE', source: 'shortcut'})
+              // pause video playback
+              videoService.send({type: 'CHANGE'})
+            },
```
Comment (vojtaholik) [inline packages/player/src/components/shortcut.tsx:133]: press `c` to focus on note input and start writing
`enter` to submit and continue playing the video

**Review by DrShpongle — COMMENTED** [review]
Just cool 👍👍

**Review by DrShpongle — APPROVED** [review]

**Pull request merged**